### PR TITLE
[Quant][FX] Lower QLinearLeakyReLU for onednn backend

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89188
* #88923
* #88879
* __->__ #88668

**Summary**
Add quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering is supported only in FX mode.

**Test plan**
python test_quantization.py TestQuantizeFx

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @leslie-fang-intel @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

File: torch/ao/quantization/backend_config/__init__.py [hunk 1]
```diff
@@ -17,4 +18,5 @@
     "BackendPatternConfig",
     "DTypeConfig",
     "ObservationType",
+    "get_onednn_backend_config",
```
Comment (jerryzh168) [inline torch/ao/quantization/backend_config/__init__.py:21]: should this change happen in previous PR?

File: test/quantization/fx/test_quantize_fx.py [hunk 2]
```diff
@@ -363,6 +365,48 @@ def forward(self, x):
...
+            m = LinearBnLeakyReluModel(with_bn).eval()
+            # fuse_fx is a top level api and only supports eval mode
+            m = fuse_fx(m,
+                        backend_config=get_onednn_backend_config())
+            self.checkGraphModuleNodes(
+                m,
+                expected_node_list=expected_nodes,
+                expected_node_occurrence=expected_occurrence)
+
+    def test_no_fuse_linear_bn_leaky_relu_eval(self):
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:391]: these two tests should be in previous PR as well I think

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 3]
```diff
@@ -253,6 +253,7 @@ def should_skip_lowering(op: torch.fx.node.Node, qconfig_map: Dict[str, QConfigA
 #   2) The replacement static quantized module class for lowering
 STATIC_LOWER_FUSED_MODULE_MAP: Dict[Type[nn.Module], Tuple[Type[nn.Module], Type[WeightedQuantizedModule]]] = {
     nni.LinearReLU: (nnqr.Linear, nniq.LinearReLU),
+    nni.LinearLeakyReLU: (nnqr.Linear, nniq.LinearLeakyReLU),
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_lower_to_native_backend.py:256]: I guess this change is related to lowering, but currently it is not tested, we should add a test that goes through PTQ flow of fx graph mode quantization and lower the linear - leaky_relu pattern, and probably another test for (linear - bn - leaky_relu)

File: torch/ao/quantization/quantization_mappings.py [hunk 4]
```diff
@@ -110,6 +110,7 @@
     nni.ConvReLU2d: nniq.ConvReLU2d,
     nni.ConvReLU3d: nniq.ConvReLU3d,
     nni.LinearReLU: nniq.LinearReLU,
+    nni.LinearLeakyReLU: nniq.LinearLeakyReLU,
```
Comment (jerryzh168) [inline torch/ao/quantization/quantization_mappings.py:113]: same here, this is in eager mode quantization flow, we should add a similar test in https://github.com/pytorch/pytorch/blob/master/test/quantization/eager/test_quantize_eager_ptq.py#L76 I think

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
this one needs some changes, please see comments inline

File: torch/ao/quantization/backend_config/__init__.py (hunk 1 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/backend_config/__init__.py:21]: OK

File: test/quantization/fx/test_quantize_fx.py (hunk 2 above)
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py:391]: OK

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 3 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:256]: I have added a test case in `test/quantization/fx/test_quantize_fx.py` as `TestQuantizeFx.test_linear_leaky_relu_lowering`

File: torch/ao/quantization/quantization_mappings.py (hunk 4 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/quantization_mappings.py:113]: We only support fusion and lowing in FX mode. I have added a test case in `test_quantize_fx.py`

**Comment by Xia-Weiwen** [comment]
> this one needs some changes, please see comments inline

Hi @jerryzh168. I have done changes per your comments. Do you have more comments? Thanks!

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168. Is it ok to land? Thanks

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 3 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:256]: do we also fuse this for fbgemm and qnnpack as well? should we skip that for these other backends?

Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:256]: No. We don't fuse this for other backends. I have added a test case to confirm the pattern is not fused by default.

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
need to discuss how lowering works for onednn, when these things are not supported in fbgemm/qnnpack

**Comment by jerryzh168** [comment]
I guess the current behavior is, if people configured to quantize linear -> leaky_relu, it will produce the quantized::linear_leaky_relu op, but when you run the model it will error out and say it is not supported in fbgemm/qnnpack, is that correct?

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 3 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:256]: For fbgemm and qnnpack, `linear + leaky_relu` are not fused and they are lowered separately. If users specify `backend='onednn'` they still need to use onednn's backend config explicitly for `prepare_fx` and `convert_fx` to fuse and lower `linear-leaky_relu`. So, a new lowering function is not needed.

**Comment by Xia-Weiwen** [comment]
> I guess the current behavior is, if people configured to quantize linear -> leaky_relu, it will produce the quantized::linear_leaky_relu op, but when you run the model it will error out and say it is not supported in fbgemm/qnnpack, is that correct?

No, linear + leaky_relu fusion is only enabled if users use onednn's backend explicitly https://github.com/pytorch/pytorch/pull/88668#discussion_r1049122284

**Comment by Xia-Weiwen** [comment]
> need to discuss how lowering works for onednn, when these things are not supported in fbgemm/qnnpack

Linear + leaky_relu are not fused for fbgemm/qnnpack

**Review by jerryzh168 — APPROVED** [review]
I see, can we add a comment to the lowering code to explain the behavior? I think there are two things
1). backend_config
2). quantized engine

I think currently we are saying if we set backend_config to onednn_backend_config, we won't see the pattern for linear - leaky_relu, but my question is that if we do have "linear - leaky_relu" pattern (e.g. if user accidently uses onednn backend_config), the current lowering code will fuse it since it is shared by all backends. and if quantized engine is set to fbgemm, we'll run the quantized::linear_leaky_relu with fbgemm backend

I think in the future, since we are asking users to explicitly specify backend_config, we should probably also ask users to explicitly call the lowering code, and we should have separate lowering code for each backend as well.

maybe a comment explaining the above would be helpful.

**Comment by Xia-Weiwen** [comment]
> I see, can we add a comment to the lowering code to explain the behavior? I think there are two things 1). backend_config 2). quantized engine
> 
> I think currently we are saying if we set backend_config to onednn_backend_config, we won't see the pattern for linear - leaky_relu, but my question is that if we do have "linear - leaky_relu" pattern (e.g. if user accidently uses onednn backend_config), the current lowering code will fuse it since it is shared by all backends. and if quantized engine is set to fbgemm, we'll run the quantized::linear_leaky_relu with fbgemm backend
> 
> I think in the future, since we are asking users to explicitly specify backend_config, we should probably also ask users to explicitly call the lowering code, and we should have separate lowering code for each backend as well.
> 
> maybe a comment explaining the above would be helpful.

I see. Shall we add the comment here https://github.com/pytorch/pytorch/blob/296713a29574e911762fd902301bf1f1ac69ab04/torch/ao/quantization/fx/_lower_to_native_backend.py#L256 or somewhere else?
BTW, do we have some mechanism to prevent end users from using a different `backend_config` as the backend they are using?

**Comment by jerryzh168** [comment]
> BTW, do we have some mechanism to prevent end users from using a different backend_config as the backend they are using?

by "backend" you are talking about quantized engine right? we don't have this mechanism right now, but we might consider adding this in the future I think

**Comment by Xia-Weiwen** [comment]
> > BTW, do we have some mechanism to prevent end users from using a different backend_config as the backend they are using?
> 
> by "backend" you are talking about quantized engine right? we don't have this mechanism right now, but we might consider adding this in the future I think

Yes, I mean qengine here.
Maybe we can choose the qengine's `backend_config` automatically behind the scenes when users do not specify `backend_config` explicitly. In that case, we do not encourage end users to set `backend_config` explicitly unless they want to do some customization. How does that sound to you?

**Comment by jerryzh168** [comment]
> > > BTW, do we have some mechanism to prevent end users from using a different backend_config as the backend they are using?
> > 
> > 
> > by "backend" you are talking about quantized engine right? we don't have this mechanism right now, but we might consider adding this in the future I think
> 
> Yes, I mean qengine here. Maybe we can choose the qengine's `backend_config` automatically behind the scenes when users do not specify `backend_config` explicitly. In that case, we do not encourage end users to set `backend_config` explicitly unless they want to do some customization. How does that sound to you?

I think we may need to store target `backend`/`qengine` in the model to do that, since qengine is used when we do inference, and `backend_config` is used when we quantize the model, we can have serialization/deserialization etc. in between. cc @vkuzo wondering what is latest on storing this information in the model?

**Comment by Xia-Weiwen** [comment]
> I think we may need to store target `backend`/`qengine` in the model to do that, since qengine is used when we do inference, and `backend_config` is used when we quantize the model, we can have serialization/deserialization etc. in between. cc @vkuzo wondering what is latest plan on storing this information in the model?

I see. Now people use `backend_config` for quantization and use `qengine` for inference. If `qengine` info is stored in the model when doing quantization, `qengine` can be selected automatically for inference thus needless to specify `qengine` explicitly. That makes sense.

**Pull request closed without merging**